sdks/python/apache_beam/io/datastore/v1/datastoreio.py

```diff
@@ -0,0 +1,289 @@
+
+    entities = sharded_queries | 'Read' >> ParDo(
+        ReadFromDatastore.ReadFn(self._project, self._datastore_namespace))
+    return entities
+
+  def display_data(self):
+    disp_data = {'project': self._project,
+                 'query': str(self._query),
+                 'num_splits': self._num_splits}
+
+    if self._datastore_namespace is not None:
+      disp_data['namespace'] = self._datastore_namespace
+
+    return disp_data
+
+  class SplitQueryFn(DoFn):
+    """A `DoFn` that splits a given query into multiple sub-queries."""
+
+    def __init__(self, project, query, namespace, num_splits):
+      super(ReadFromDatastore.SplitQueryFn, self).__init__()
+      self._datastore = None
+      self._project = project
+      # using _namespace conflicts with DisplayData._namespace
+      self._datastore_namespace = namespace
+      self._query = query
+      self._num_splits = num_splits
+
+    def start_bundle(self, context):
+      self._datastore = helper.get_datastore(self._project)
+
+    def process(self, p_context, *args, **kwargs):
+      # random key to be used to group query splits.
+      key = 1
+      query = p_context.element
+
+      # If query has a user set limit, then the query cannot be split.
+      if query.HasField('limit'):
+        return [(key, query)]
+
+      # Compute the estimated numSplits if not specified by the user.
+      if self._num_splits <= 0:
+        estimated_num_splits = ReadFromDatastore.get_estimated_num_splits(
+            self._project, self._datastore_namespace, self._query,
+            self._datastore)
+      else:
+        estimated_num_splits = self._num_splits
+
+      logging.info("Splitting the query into %d splits", estimated_num_splits)
+      try:
+        query_splits = query_splitter.get_splits(
+            self._datastore, query, estimated_num_splits,
+            helper.make_partition(self._project, self._datastore_namespace))
+      except Exception:
+        logging.warning("Unable to parallelize the given query: %s", query,
+                        exc_info=True)
+        query_splits = [(key, query)]
+
+      sharded_query_splits = []
+      for split_query in query_splits:
+        sharded_query_splits.append((key, split_query))
+        key += 1
+
+      return sharded_query_splits
+
+    def display_data(self):
+      disp_data = {'project': self._project,
+                   'query': str(self._query),
+                   'num_splits': self._num_splits}
+
+      if self._datastore_namespace is not None:
+        disp_data['namespace'] = self._datastore_namespace
+
+      return disp_data
+
+  class ReadFn(DoFn):
+    """A DoFn that reads entities from Cloud Datastore, for a given query."""
+
+    def __init__(self, project, namespace=None):
+      super(ReadFromDatastore.ReadFn, self).__init__()
+      self._project = project
+      # using _namespace conflicts with DisplayData._namespace
```

Comment:
(do we really need this comment everywhere in the file.) feel free to ignore this comment.